lib/tesla/middleware/digest_auth.ex

```diff
@@ -0,0 +1,110 @@
+defmodule Tesla.Middleware.DigestAuth do
+  @moduledoc """
+  Digest access authentication middleware
+
+  [Wiki on the topic](https://en.wikipedia.org/wiki/Digest_access_authentication)
+
+  Currently the implementation is incomplete and works only for MD5 algorithm and
+  auth qop.
+
+  Example:
+      defmodule MyClient do
+        use Tesla
+
+        def client(username, password, opts \\ %{}) do
+          Tesla.build_client [
+            {Tesla.Middleware.DigestAuth, Map.merge(%{username: username, password: password}, opts)}
+          ]
+        end
+      end
+
+  Options:
+  - `:username`  - username (defaults to `""`)
+  - `:password`  - password (defaults to `""`)
+  - `:cnonce_fn` - custom function generating client nonce (defaults to `&Tesla.Middleware.DigestAuth.cnonce/0`)
+  - `:nc`        - nonce counter (defaults to `"00000000"`)
+  """
+
+  def call(env, next, opts) do
+    if Keyword.get(env.opts, :digest_auth_handshake) do
```

Comment:
Yep - https://github.com/teamon/tesla/blob/master/lib/tesla.ex#L128

(We need a way to change that, but e.g. BaseUrl middleware uses simple string instead of opts, and Headers also requires a map)